### DISCUSSION TITLE
What is difference between jaeger collector and jaeger agent?

### DISCUSSION THREAD

**Discussion by zheyiyi** [jaegertracing/jaeger#1488] [Q&A] [answered]
Why do you directly send trace spans to jaeger collector from jaeger client? What does jaeger agent do?

**Comment by yurishkuro** [comment, marked as answer]
Agents are local to the host, so the client libraries do not need to do any service discovery to find them. Agents mostly just forward traffic to collectors, but they can also enrich spans with extra metadata (using agent tags).

~For legacy reasons, agents implement HTTP endpoints that clients use to poll for sampling strategies and other configurations. The same HTTP endpoints are not mounted on the collectors.~ (this has been fixed)

**Comment by teivah** [comment]
@yurishkuro Just an additional question, please. So why do we need to specify the collector endpoint in each client? I mean if the traffic is client -> agent -> collector, why should a client be aware of the collector endpoint?
Thanks

**Comment by yurishkuro** [comment]
you don't need to specify collector endpoint. If you do, the client will talk directly to collector, bypassing the agent.

**Reply by vikaskarla** [reply]
Would Jaeger be able to send traces to a third party collector in zipkin V2 format? I have a use-case like this where I am trying to use Jaeger(in Python) to generate traces and transport the traces using HTTP to a third party trace collector service. Thanks!

**Comment by a1300** [comment]
FYI this is also documented in the [FAQ](https://www.jaegertracing.io/docs/1.21/faq/) on the documentation website

**Comment by kinthaiofficial** [comment]
Good question — the architecture distinction matters a lot for reliability planning.

**Jaeger Agent** (deprecated in newer versions) ran as a sidecar, receiving spans via UDP from your app and forwarding to the Collector. The UDP part meant fire-and-forget — fast, but spans could silently drop under load.

**Jaeger Collector** receives spans directly (HTTP/gRPC), validates, and writes to storage. It's stateless and horizontally scalable.

The shift in newer Jaeger versions is to go direct-to-collector via OTLP, skipping the agent entirely:

```yaml
# Old pattern (via agent sidecar)
JAEGER_AGENT_HOST: jaeger-agent
JAEGER_AGENT_PORT: 6831  # UDP

# New pattern (direct to collector)
OTEL_EXPORTER_OTLP_ENDPOINT: http://jaeger-collector:4317
OTEL_EXPORTER_OTLP_PROTOCOL: grpc
```

For multi-service systems the direct OTLP approach is cleaner — you get backpressure signaling (gRPC flow control) instead of silent UDP drops. We use this in a setup where multiple AI agents all export traces to the same collector, and being able to detect "collector overloaded" vs "spans lost" makes debugging much easier.

Related write-up on tracing in distributed agent coordination: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons
